```{r librairies-systematicAnalysis}
require(tidyverse)
require(ggpubr)
require(magrittr)
require(ggimage)
require(ggrepel)

tibble(a=c("a", "b", "c", "a"), b=c("a","d", "e", "f")) %>% distinct()
```


```{r parameters-systematicAnalysis}
proteomicRanking.cutoff <- 1800
```

```{r importDatasets-SA, include=F, cache=T}
#import from package data, data wrangling scripts are in data-raw
t_transcriptomicZhao2018 <- CveklRBP2019::t_transcriptomicZhao2018
t_proteomicZhao2018 <- CveklRBP2019::t_proteomicZhao2018 %>%
  rename(gene=GeneNames)
```

```{r filterRBP-SA, cache=T}
#import proteomic dataset
t_proteomicZhao2018_RBP <- read_rds("../data/derived_data/t_proteomicZhao2018_RBP.rds")
```

```{r functionalanalysis}
###########################################################
#import DAVID annotations
###########################################################
t_DAVID_annotationsRBP <- read_tsv("../data/raw_data/functionalanalysis/DAVID_annotations.tsv") %>%
  #select all annotations columns
  dplyr::select(ID, UP_KEYWORDS, starts_with("GO"), starts_with("KEGG")) %>%
  #gather all columns in one
  unite(col = "KEYWORDS", UP_KEYWORDS, KEGG_PATHWAY, starts_with("GOTERM"), sep="") %>%
  #transform table so that one row = one gene and one keyword
  separate_rows(KEYWORDS, sep=",") %>%
  #clean table
  dplyr::filter(!KEYWORDS=="") %>% drop_na() %>%
  mutate(KEYWORDS=str_replace_all(KEYWORDS, "GO:", "GO")) %>%
  separate(col="KEYWORDS", into=c("keyID", "keyTerm"), sep="~|:", fill = "left") %>% 
  dplyr::select(gene=ID, keyTerm) %>%
  write_tsv("../data/derived_data/DAVID_annotationsRBP_processed.tsv") 

###########################################################
#import Home annotations
###########################################################

t_CVEKL_annotationsRBP <- read_tsv("../data/raw_data/functionalanalysis/CVEKL_annotationsRBP.tsv",
  col_types = "cccc",
  col_names = c("gene", "groups", "subgroups", "category"),
  skip=1)  %>% 
  unite(col="keyTerm", groups:category, sep=", ") %>%
  separate_rows(keyTerm, sep=", ")  %>% 
  mutate(keyTerm = case_when(keyTerm=="splicing" ~ "mRNA splicing", 
                             TRUE ~ keyTerm)) %T>%
  write_tsv("../data/derived_data/CVEKL_annotationsRBP_processed.tsv") 


###########################################################
#compile annotations
###########################################################
t_functionalGridRBP <- read_tsv("../data/raw_data/functionalanalysis/RBP_functionalannotationsGrid.tsv")

t_annotationsRBP <- bind_rows(t_DAVID_annotationsRBP, t_CVEKL_annotationsRBP) %>%
  #process our annotations table through the grid
  #note that annotations that are not represented are dropped
  right_join(t_functionalGridRBP, by="keyTerm") 


###########################################################
#post-process to establish a hierarchy of annotations
###########################################################


t_annotationsRBP <- t_annotationsRBP %>%
  mutate(category = as.character(category),
         category = case_when(
           category=="pre-mRNA processing and transport" ~ "pre-mrna processing",
           TRUE ~ category
         ))  %>%
  mutate(group1 = case_when(grepl("putative", category, ignore.case = T) ~ "putative RBP",
                            grepl("putative", category, ignore.case = T) == FALSE ~ "RBP"),
         group5 = case_when(grepl("translational regulation", category, ignore.case = T) ~ "translational regulation",
                            grepl("translational initiation", category, ignore.case = T) ~ "translational initiation",
                            grepl("ribosome", category, ignore.case = T) ~ "ribosome"),
         #grepl("splice", keyTerm, ignore.case = T) ~ "Spliceosomal complex",
         #grepl("alternative splicing", category, ignore.case = T) ~ "Alternative splicing"),
         group4 = case_when(grepl("stress", group,ignore.case = T) ~ "Stress response",
                            grepl("translation|ribosome", category, ignore.case = T) ~ "Translation",
                            grepl("mRNA transport", category, ignore.case = T) ~ "mRNA transport",
                            grepl("mRNA stability", category, ignore.case = T) ~ "mRNA stability",
                            grepl("polyadenylation|decapping", keyTerm, ignore.case = T) ~ "mRNA processing",
                            grepl("mRNA processing", category, ignore.case = F) ~ "mRNA processing",
                            grepl("mRNA sequestration", category, ignore.case = T) ~ "mRNA sequestration",
                            category == "pre-mrna processing" ~ "pre-mRNA Transport and processing",
                            grepl("spliceo", category, ignore.case = T) ~ "splicing",
                            keyTerm == "mRNA splicing" ~ "splicing",
                            keyTerm == "Alternative splicing" & keyTerm == "splicing" ~ "splicing",
                            grepl("mRNA degradation", category) ~ "mRNA degradation",
                            grepl("snRN", keyTerm, ignore.case = T) ~ "snRNA binding",
                            grepl("tRNA", keyTerm, ignore.case = T) ~ "tRNA binding",
                            grepl("rRNA", keyTerm, ignore.case = T) ~ "rRNA binding"
         ),
         group3 = case_when(group == "ncRNA binding" ~ "ncRNA binding",
                            group == "mRNA binding" ~ "mRNA binding",
                            group == "pre-mRNA binding" ~ "pre-mRNA binding",
                            grepl("Stress response|Translation|mRNA transport|mRNA stability|mRNA sequestration|mRNA processing|mRNA degradation", group4) ~ "mRNA binding",
                            grepl("snRNA|tRNA|rRNA", group4) ~ "ncRNA binding",
                            grepl("pre-mRNA|splicing|alternative splicing", group4) ~ "pre-mRNA binding")) 

list.putativeRBP <- t_annotationsRBP %>% dplyr::filter(category=="putative RBP") %$% gene

t_annotationsRBP_processed <- t_annotationsRBP %>% 
  dplyr::filter(!gene%in%list.putativeRBP) %>% #remove putative RBP
  dplyr::select(gene, group1, group3, group4, group5) %>% 
  drop_na(gene) %>% 
  ungroup() %>% 
  {mutate(., na_count = apply(., 1, function(x) sum(is.na(x))))} %>%
  group_by(gene) %>% 
  #putative RBP must not overlap with RBP annotations
  dplyr::filter(group1==group1[nchar(group1)==max(nchar(group1))]) %>% ungroup() %>% 
  distinct() %T>%
  write_tsv("../data/derived_data/RBP_functionalannotations_processed.tsv")


t_annotationsRBP_annotated <- t_annotationsRBP %>%
  group_by(gene, group3, group4, group5) %>% summarise()  %>% 
  right_join(t_proteomicZhao2018_RBP, by="gene")  %>%
  gather(group3, group4, group5, key="level", value="group") %>%
  dplyr::filter(!is.na(group)) %T>%
  write_tsv("../data/derived_data/RBP_functionalannotations_annotated.tsv")

```



```{r figSystematic, include=T, echo=FALSE, warning=F, message=F, fig.cap="\\textbf{\\label{fig:articlevennRBP}Systematic analysis of lens-expressed RBPs.} (A) Venn diagram of highly-expressed RBPs grouped in either lens epithelial or fiber cells or both. (B) Functional representativity of lens-expressed RBPs, divided with regard of target RNAs class. (C) Dotplot showing the functional classification of expressed RBPs in the lens. The tissue specificity of an RBP is indicated by a colored sign next to its name.", fig.height=9, fig.width=7}

fig_barplotSA <- source("../figures/barplotSA.R")
fig_functionalSA <- source("../figures/functionalSA.R")

p <- ggplot() + theme_void() + theme(aspect.ratio = 0.6875)
q <- ggplot() + theme_void() + theme(aspect.ratio = 0.5)


q1 <- ggbackground(p, "../figures/vennDiagram_redblue.png") + 
  theme(aspect.ratio = 0.6875, plot.margin = unit(rep(.4,4), "lines"))

#pdf("../figures/systematicAnalysisRBP.pdf", width = 6.3, height = 7.9)
ggarrange(ggarrange(q1,fig_barplotSA$value, ncol=2, nrow=1, widths = c(.3,.7), labels = c("A", "B")), fig_functionalSA$value , nrow=2, heights =c(.3,.7), labels = c("", "C")) 
#dev.off()
```

```{r functionalFigure, include=T, fig.width=6.29, fig.height=2.9} 

t_ISYTE_enrichment <- read_tsv("../data/raw_data/functionalanalysis/ISYTE_enrichment.tsv") %>% dplyr::rename(gene=Symbol)

t_proteomicEnrichment <- t_ISYTE_enrichment %>%
  gather(-gene, -Rank, key="stage", value="enrichement") %>%
  mutate(stage=str_extract(string = stage, "E[0-9]{1,2}|P[0-9]{1,2}"), #extract stage
         enrichement = as.numeric(enrichement)) %>%
  group_by(gene) %>%
  summarise(enrich_category = max(enrichement, na.rm = T)) %>%
  mutate(enrich_category = case_when(enrich_category < 2 ~ "< 2",
                                     enrich_category >= 2 & enrich_category < 5 ~ "[2;5[",
                                     enrich_category >= 5 & enrich_category < 10 ~ "[5;10[",
                                     enrich_category >= 10 & enrich_category < 50 ~ "[10;50[",
                                     enrich_category >= 50 ~ "> 50")) %>%
  mutate(enrich_category = factor(enrich_category, levels = c("< 2", "[2;5[", "[5;10[","[10;50[", "> 50")))%>%
  right_join(t_annotationsRBP_figure, by="gene") %>% 
  filter(groups %in% c("translational regulation", "mRNA degradation", "mRNA stability")) %>%
  mutate(groups = factor(groups, levels = c("mRNA stability", "translational regulation","mRNA degradation"))) %>%
  replace_na(list(enrich_category="NA"))

listRBPinterest <- "Caprin2,Carhsp1,Celf1,Ddx39b,Fxr1,Igf2bp1,Igf2bp3,Hnrnpd,Rbm8a,Lsm5,Magohb,Pabpc1,Ptbp1,Tardbp,Rbm38,Syncrip,Tdrd7,Ybx1,Ybx3,Lsm2,Eif5a,Pdcd4,Upf1" %>% 
  str_split(pattern = ",") %>% 
  unlist()

t_proteomicEnrichment_interest <- t_proteomicEnrichment %>% 
  filter(gene %in% listRBPinterest) 

t_proteomicEnrichment %>%
  ggplot(aes(x=rankEpith, y=rankFibers, color=tissue, size=enrich_category, shape=tissue)) +
  geom_point() + 
  geom_point(data = filter(t_proteomicEnrichment_interest, tissue=="both"), 
             mapping = aes(x=rankEpith, y=rankFibers, size=enrich_category), 
             color="#c0c0c0", shape=16, inherit.aes = F, show.legend = F) +
  geom_point(data=filter(t_proteomicEnrichment_interest, tissue=="epi"), 
             mapping = aes(x=rankEpith, y=rankFibers, size=enrich_category), 
             color="#0000ff", shape=17, inherit.aes = F, show.legend = F) +
  geom_point(data=filter(t_proteomicEnrichment_interest, tissue=="fibers"), 
             mapping = aes(x=rankEpith, y=rankFibers, size=enrich_category), 
             color="#ff0000", shape=15, inherit.aes = F, show.legend = F) +
  scale_size_discrete(range= c(1, 4)) +
  xlim(c(0,NA))+ ylim(c(0,NA))+
  scale_shape_manual(values=c(1, 2, 0), 
                     labels=c("Both", "Lens epithelium", "Lens fibers")) +
 geom_text_repel(data = t_proteomicEnrichment_interest, 
                 mapping = aes(label=gene, x=rankEpith, y=rankFibers), 
                 inherit.aes = F,
                 fontface = 2, 
                 color = "black", 
                 segment.color = "black",
                 size = 2.4, 
                 box.padding =  unit(.2, "lines"), 
                 segment.size = .4, 
                 min.segment.length = unit(1, "mm"),
                 show.legend = F, 
                 max.iter = 10000, 
                 seed = 22, 
                 force = 3) + 
  scale_color_manual(values = c("epi"="#0000ff", 
                                 "fibers" = "#ff0000",
                                 "both" = "#c0c0c0"), 
                     labels=c("Both","Lens epithelium", "Lens fibers")) +
  geom_hline(yintercept=1800, color="black", lty="dotted") + 
  geom_vline(xintercept=1800, lty="dotted")+ 
  guides(size = guide_legend(title = "Maximum enrichment in the lens (iSyTe)", title.position = "top", title.hjust = .5), 
         color=guide_legend(title="proteomic rank < 1800", title.position = "top", title.hjust = .5),
         shape=guide_legend(title="proteomic rank < 1800", title.position = "top", title.hjust = .5))+  theme_gray() + 
  ylab("proteomic rank in lens fiber cells")+ 
  xlab("proteomic rank in lens epithelial cells") +
  theme(panel.border = element_rect(color="black", fill = "transparent"), 
        axis.text.x = element_text(size=7, color="black"),
        axis.text.y = element_text(size=7, color="black", angle=90, vjust=1, hjust=.5),
        axis.title = element_text(size=7),
        axis.ticks = element_line(size=.3, color="black"),
        strip.background = element_rect(color="black", fill = "white"),
        panel.background = element_rect(fill="transparent"),
        panel.grid = element_blank(),
        axis.text = element_text(size=8, color="black"),
        strip.text = element_text(face="bold", size=7), 
        legend.position = "top",
        legend.margin = margin(0,0,0,10, "mm"),
        legend.spacing = unit(0, "cm"),
        legend.direction = "horizontal",
        legend.box = "horizontal",
        legend.title = element_text(size=7, face = "bold"), 
        legend.box.margin = margin(0,0,0,0,"mm"),
        legend.key = element_blank(),
        legend.text = element_text(size=7)) + 
  facet_wrap(~groups, scales = "free") + 
  theme(plot.margin = margin(.05,.12,.05,.05,"cm"))
```


